### ISSUE TITLE
Add lint for u8 as *mut cast

### ISSUE THREAD

**Issue by mattico** [rust-lang/rust#42915]
This is an easy error to make: #42901 #42827

A cast from u8 to a pointer is probably always an error. In the rare case that you're creating a pointer into the first 256 bytes of address space, `u8 as usize as *mut` is more clear, or you can be troubled to `#[allow(u8_to_ptr)]`.

**Comment by retep998** [comment]
This lint should apply to any cast from an integer type that is not `usize/isize` to a raw pointer, especially because Rust has awful footgun behavior for such casts.

`-1i8 as isize` gives `0xFFFFFFFFFFFFFFFF`
`-1i8 as usize` gives `0xFFFFFFFFFFFFFFFF`
`-1i8 as *const () as usize` rather surprisingly gives `0xFF`

If we compare this to C++ for example:

`(intptr_t)int8_t{-1}` gives `0xFFFFFFFFFFFFFFFF`
`(uintptr_t)int8_t{-1}` gives `0xFFFFFFFFFFFFFFFF`
`(uintptr_t)(void*)int8_t{-1}` gives the naturally expected `0xFFFFFFFFFFFFFFFF`

So basically Rust has a giant footgun in that casting from non pointer sized integers to pointers differs from C/C++ and casts from non pointer sized integers to pointer sized integers. This lint is *essential* to catch this footgun and warn people about this hazard!

**Comment by retep998** [comment]
Related footgun: https://github.com/rust-lang/rust/issues/43291

**Comment by Enselic** [comment]
Triage: Lint remains unimplemented. Seems useful to have and easy to implement. Maybe first in clippy? Not sure. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=23a52681da69cc3e5ba5b5071765b2c7): 

```rust
fn main() {
    let _ = 42u8 as *mut u8;
}
```
